index.js

```diff
@@ -83,182 +86,301 @@ function fastifyOauth2 (fastify, options, next) {
   }
 
-  function clearCodeVerifierCookie (reply) {
-    reply.clearCookie(VERIFIER_COOKIE_NAME, cookieOpts)
+  if (discovery) {
+    discoverMetadata(discovery.issuer, (err, fetchedMetadata) => {
+      if (err) {
+        next(err)
+        return
+      }
+      const authFromMetadata = getAuthFromMetadata(fetchedMetadata)
+
+      const discoveredOptions = {
+        ...options,
+        credentials: {
+          ...options.credentials,
+          auth: authFromMetadata
+        }
+      }
+      // respect users choice if they provided PKCE method explicitly
+      // even with usage of discovery
+      if (!options.pkce) {
+        // otherwise select optimal pkce method for them,
+        // if such can be found in metadata response, ofcourse
+        const optimalPkceMethod = selectPkceFromMetadata(fetchedMetadata)
+        if (optimalPkceMethod) {
+          discoveredOptions.pkce = optimalPkceMethod
+        }
+      }
+      configure(discoveredOptions)
+      next()
+    })
+  } else {
+    configure(options)
+    next()
   }
+  function discoverMetadata (issuer, cb) {
+    const discoveryUri = getDiscoveryUri(issuer)
+
+    const httpOpts = {
+      headers: {
+        ...options.credentials.http?.headers,
+        'User-Agent': userAgent
+      }
+    }
+    if (omitUserAgent) {
+      delete httpOpts.headers['User-Agent']
+    }
 
-  const oauth2 = new AuthorizationCode(credentials)
+    const req = (discoveryUri.startsWith('https://') ? https : http).get(discoveryUri, httpOpts, onDiscoveryResponse)
 
-  if (startRedirectPath) {
-    fastify.get(startRedirectPath, { schema }, startRedirectHandler)
+    req.on('error', (e) => {
+      const err = new Error('Problem calling discovery endpoint. See innerError for details.')
+      err.innerError = e
+      cb(err)
+    })
+
+    function onDiscoveryResponse (res) {
+      let rawData = ''
+      res.on('data', (chunk) => { rawData += chunk })
+      res.on('end', () => {
+        try {
+          cb(null, JSON.parse(rawData))
+        } catch (err) {
+          cb(err)
+        }
+      })
+    }
   }
+}
 
-  const decoration = {
-    oauth2,
-    getAccessTokenFromAuthorizationCodeFlow,
-    getNewAccessTokenUsingRefreshToken,
-    generateAuthorizationUri,
-    revokeToken,
-    revokeAllToken
+function getDiscoveryUri (issuer) {
+  // eslint-disable-next-line
+  const parsed = url.parse(issuer)
+
+  if (parsed.pathname.includes('/.well-known/')) {
+    return issuer
+  } else {
+    let pathname
+    if (parsed.pathname.endsWith('/')) {
+      pathname = `${parsed.pathname}.well-known/openid-configuration`
+    } else {
+      pathname = `${parsed.pathname}/.well-known/openid-configuration`
+    }
+    return url.format({ ...parsed, pathname })
   }
+}
 
-  try {
-    fastify.decorate(name, decoration)
-    fastify.decorate(`oauth2${name.slice(0, 1).toUpperCase()}${name.slice(1)}`, decoration)
-  } catch (e) {
-    next(e)
-    return
+function selectPkceFromMetadata (metadata) {
```

Comment:
Always use S256 PKCE even when code_challenge_methods_supported is not given.

Reply:
AS should ignore it if doesn't support it, right?